java/ql/src/Telemetry/AutomodelSharedCharacteristics.qll

```diff
@@ -0,0 +1,281 @@
+float maximalConfidence() { result = 1.0 }
+
+float highConfidence() { result = 0.9 }
+
+float mediumConfidence() { result = 0.6 }
+
+/**
+ * A specification of how to  instantiate the shared characteristics for a given candidate class.
+ *
+ * The `CandidateSig` implementation specifies a type to use for Endpoints (eg., `ParameterNode`), as well as a type
+ * to label endpoint classes (the `EndpointType`). One of the endpoint classes needs to be a 'negative' class, meaning
+ *   "not any of the other known endpoint types".
+ */
+signature module CandidateSig {
+  class Endpoint;
+
+  class EndpointType;
+
+  /** Gets the string representing the file+range of the endpoint. */
+  string getLocationString(Endpoint e);
+
+  /**
+   * Defines what labels are known, and what endpoint type they correspond to.
+   */
+  predicate isKnownLabel(string label, string humanReadableLabel, EndpointType type);
+
+  /**
+   * EndpointType must have a 'negative' type that denotes the absence of any sink.
+   * This predicate should hold for that type, and that type only.
+   */
+  predicate isNegative(EndpointType t);
```

Comment:
If it's a requirement that we have a negative endpoint type, should we make this a `class NegativeEndpointType extends EndpointType` defined in the signature? A predicate means the instantiation could just implement as `none()`.

Reply:
1) I really like this, and it's also more similar to what a similar use case elsewhere does ([codeql/shared/regex/codeql/regex/RegexTreeView.qll](https://github.com/github/codeql/blob/bd3aaf0306d4e0382ae1cda051a63926f0a6dee3/shared/regex/codeql/regex/RegexTreeView.qll#L13))
2) Even though I'll build this as suggested, for completness' sake: it still doesn't solve the `none()` problem (eg., the instantiation could use something like `class Neg extends EndpointType { Neg() { none() } }`